sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc

```diff
@@ -0,0 +1,1191 @@
+
+The API for explicitly adding nodes to a `command_graph` includes a
+`property_list` parameter. This extension defines the `depends_on` property to
+be passed here. `depends_on` defines any `node` objects for the created node to
+be dependent on, and therefore form an edge with. These nodes are in addition to
+the dependent nodes identified from the command-group requisites of the created
+node.
+
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental::property::node
+class depends_on {
+  public:
+    template<typename... NodeTN>
+    depends_on(NodeTN... nodes);
+};
+}
+----
+
+=== Graph
+
+This extension adds a new `command_graph` object which follows the
+{crs}[common reference semantics] of other SYCL runtime objects.
+
+A `command_graph` represents a directed acyclic graph of nodes, where each node
+represents a single command or a sub-graph. The execution of a graph completes
+when all of its nodes have completed.
+
+A `command_graph` is built up by either recording queue submissions or
+explicitly adding nodes, then once the user is happy that the graph is complete,
+the graph instance is finalized into an executable variant which can have no
+more nodes added to it. Finalization may be a computationally expensive
+operation as the runtime is able to perform optimizations based on the graph
+structure. After finalization the graph can be submitted for execution on a
+queue one or more times with reduced overhead.
+
+==== Graph State
+
+An instance of a `command_graph` object can be in one of two states:
+
+* **Modifiable** - Graph is under construction and new nodes may be added to it.
+* **Executable** - Graph topology is fixed after finalization and graph is ready to
+  be submitted for execution.
+
+A `command_graph` object is constructed in the _recording_ state and is made
+_executable_ by the user invoking `command_graph::finalize()` to create a
+new executable instance of the graph. An executable graph cannot be converted
+to a modifiable graph. After finalizing a graph in the modifiable state it is
+valid for a user to add additional nodes and finalize again to create subsequent
+executable graphs. The state of a `command_graph` object is made explicit by
+templating on state to make the class strongly typed, with the default template
+argument being `graph_state::modifiable` to reduce code verbosity on
+construction.
+
+.Graph State Diagram
+[source, mermaid]
+....
+graph LR
+    Modifiable -->|Finalize| Executable
+....
+
+==== Executable Graph Update
+
+A graph in the executable state can have each nodes inputs & outputs updated
+using the `command_graph::update()` method. This takes a graph in the
+modifiable state and updates the executable graph to use the node input &
+outputs of the modifiable graph, a technique called _Whole Graph Update_. The
+modifiable graph must have the same topology as the graph originally used to
+create the executable graphs, with the nodes added in the same order.
```

Comment:
I don't understand how "graph update" is supposed to work.  I think my confusion is with this statement: "The modifiable graph must have the same topology as the graph originally used to create the executable graphs".  What precisely is the "topology"?  Clearly, this means that the number of nodes and the dependencies between the nodes cannot change.  Is it allowed to change the command inside each node?  To me that would also be part of the topology, so I would not expect it to be possible to change the commands in each node either.

But then what can you change?  If you can't change the number of nodes, their dependencies, or the command in each node, what's left?

Reply:
I see now.  I wonder if we could expose this more clearly by adding a member function `change_parameter` like this:

```
template<>
class command_graph<graph_state::executable> {
public:
    command_graph() = delete;
    change_parameter(void *from, void *to);
};
```

This would work in the case where the graph used USM pointers for input / output.  Let's say the graph was originally created to use USM pointer `x` in some kernel(s), where `x` was allocated outside of the graph.  The user could call `change_parameter(x, y)` and resubmit the graph.  The second graph submission would run the kernel(s) using `y` instead of `x`.

In theory, there could be an overload of `change_parameter` that changes a `buffer` that is used by the kernels.  This would require you to track down all accessors that use the buffer.

Reply:
I think your proposed interface is conceptually easier to understand, my only concern with it is that it requires the application to do bookkeeping to track `from`, especially over multiple
iterations of update. Another more tenuous point is that in a hypothetical future where you could serialize/deserialize an executable graph from an offline compiled graph, there would be no `from`.

We could provide both the `change_parameter` and whole graph update interfaces, but that might not achieve making whole graph update any clearer.

Reply:
> Another more tenuous point is that in a hypothetical future where you could serialize/deserialize an executable graph from an offline compiled graph, there would be no from.

You are considering some future API where the user can store a graph as a file (say) and then read it back into another process?  If you add an API like this, don't you need to somehow recover the "from" parameters anyway?  Otherwise, what would you pass to the graph once it was executed after being read from storage?

I'm not exactly sure how you would do this.  Maybe the user would need to supply the "from" values when reading the graph from storage.  My point is only that you will need to recover the "from" arguments anyways, so the `change_parameter` API would still make sense.

Reply:
> You are considering some future API where the user can store a graph as a file (say) and then read it back into another process? 

Indeed, that was one of the long term ideas we got feedback that the graphs concept would enable. But you are right, we wouldn't have a modifiable graph anyway in this scenario to do a whole graph update from. Thinking now that some sort of `get_parameters()` query might be nice to complement `change_parameters()`. 

I think a `change_parameters()` API should also be able to set multiple new parameters at once. Setting them individually is inefficient when a user wants to update N inputs/outputs the graph between executions (especially if a runtime emulates this is in software). That's just a matter of making the signature `change_parameters(std::vector<void*> from, std::vector<void*> to)` or something equivalent though (maybe a vector of `std::pair` instead).

The other benefit of whole graph update is that it could work when you are using record & replay with an external library call to create a graph. Then you could update the graph without knowing exactly the SYCL buffer/USM parameters you're updating. e.g. something like:

```cpp
command_graph<modifiable> graph;

int inputs[4] = {1,2,3,4}; // Used to inform library input/outputs
graph.begin_recording(q);
library_call(q, inputs[0]); 
graph.end_recording();

auto exec_graph = graph.finalize(q.get_context());
q.ext_oneapi_graph(exec_graph);
for (i=1; i < 4; i++) {
  command_graph<modifiable> graph;
  graph.begin_recording(q);
  library_call(q, inputs[i]);
  graph.end_recording();
  exec_graph.update(graph);  // could be faster than creating a new executable graph.
  q.ext_oneapi_graph(exec_graph);
}
```

Having both interfaces is probably my preference to cover the most usecases, but `set_parameters` is the easier to understand and use interface. So we could start with only that interface in the extension. Then add whole graph update back in if both a) record & replay is validated as being a viable way to create a graph (from your earlier comment suspecting it to be error prone), and b) we see the above library recording use case in real applications we want to support.

Reply:
Is there any chance that modifying a parameter from value1 (currently set) to value2 (desired) is a bad idea?

Your vector of parameters looks a little bit like an INFO object In MPI (a set of {key, value} pairs). MPI permits the implementation to ignore changes when they are inconvenient -- for example, doing the change requested would mean recreating the object from scratch but the object is in use elsewhere so destroying it is bad.

In MPI, we have the semantic of requested/provided, MPI_COMM_SET_INFO is an attempt to set all the info keys to the values given in the input INFO object argument, but there is an expectation that some will be accepted and some will not. The provided portion involves a subsequent call to MPI_COMM_GET_INFO, which returns all {key, value} that are (now) set. The user must compare the input INFO with the output INFO to figure what actually happened.

The rules might be different here for an executable graph than for a modifiable graph?

MPI mostly only has the equivalent of executable (e.g. communicators/windows/files/requests are all hard to change once they exist). It has the modifiable-until-made-executable for MPI_Datatype objects, but it does not make much use of the semantic difference because MPI_Datatype objects are immutable: any change creates a new one.

Perhaps when "change parameter(s)" is applied to an executable graph you want it to produce a new executable graph that has the "to" parameters set? Internally, it might re-use most of the "from" graph, but the user would see executable graph objects as immutable.

Considerations like immutability are especially important in multi-threaded scenarios where multiple threads might be concurrently interacting with the same executable graph.

Reply:
> Thinking now that some sort of `get_parameters()` query might be nice to complement `change_parameters()`.

Yes, that might be nice.  What other queries would be useful for parameters?  Maybe:

* Get the number of parameters
* Query the type of each parameter (USM pointer vs. accessor)?

> I think a `change_parameters()` API should also be able to set multiple new parameters at once

It's very convenient to have APIs that build things incrementally, so I'm not sure you really need a single function that can set all the parameters at once.  There are other ways to address the overhead.  I was assuming that `change_parameter` would actually just cache some information without rewriting the graph, so the call should be very fast.  The overhead of rewriting the graph could be done in either of two ways.

Option 1: Add an `update` function.  In this option, the user first changes parameters via `change_parameter` but the graph doesn't use the new parameter until `update` is called.  The `update` function could even return a new graph object:

```
template<>
class command_graph<graph_state::executable> {
public:
    command_graph() = delete;
    change_parameter(void *from, void *to);
    command_graph update();
};
```

Option 2: Simply cache parameter changes in `change_parameter` but don't rewrite the graph until it is submitted to a queue.  At this point, you check to see if there are any cached parameter changes and then rewrite the graph if there are.

Reply:
Caching the changes pushes in the opposite direction to the "give timely and meaningful errors" concern. The `update` call would need to be able to list all parameter changes that caused problems, which is complex structured information. Having `change_parameter` give an error would be more timely and meaningful, but requires more than "just cache it".

Reply:
> Having change_parameter give an error would be more timely and meaningful, but requires more than "just cache it".

What sort of errors can arise from changing a parameter?

Reply:
> What sort of errors can arise from changing a parameter?

"That is the right question" (Dr Lanning's hologram, I Robot) -- I don't know, what parameters exist or will exist?

For MPI, setting some INFO keys to some values is hard/impossible once the object exists because it needs an important internal data structure to be recreated in an incompatible manner. A new object of the same type with that key set to that value is more likely to be possible/easy. This happens when the INFO key/value pair relaxes a rule that is normally enforced -- re-imposing the rule retrospectively on already in-flight operations is not possible because there is insufficient information to reverse engineer the state the object would have been in had the rule been enforced.

An example from MPI is the non-overtaking rule -- normally, messages do not overtake each other; setting mpi_allow_overtaking to true, starting some messages, then trying to set it back to false is a problem because MPI might have allowed messages to overtake and may not have kept enough information about the original order to reconstruct that order. Thus, setting mpi_allow_overtaking to false can result in a "computer says no" response.

Reply:
I think you are misinterpreting what the word "parameter" means in this context.  When we say "parameter" here, we are talking about a USM pointer or a SYCL accessor that kernels in the graph use to access memory.  Therefore "change parameter" means "create a new graph that accesses memory from pointer B instead of pointer A".

Reply:
I think that being able to query the number and type (and potentially other information) certainly makes sense. In that scenario though I think that a get/set approach would be preferable to decouple the `from` and the `to` and take the place of `change_parameter`. In that case it may be better to associate each parameter with an ID that can be used when setting parameters rather than requiring the user to maintain an up-to-date record of which pointer value was previously associated with the graph.

The idea of batching individual changes into an `update` function could be of benefit, though I don't believe it should return a new executable graph. The idea of updating an existing executable graph is that the user does not wish to pay the finalization cost, and that by updating you may (depending on backend capability) be able to avoid having to pay that by instead directly updating the parameters of the backend structures rather than re-creating them. Creating a new executable graph would force us to recreate everything for the new graph removing the benefit of updating this way. We also have restrictions on graphs which use allocation nodes in that only one executable graph created from a single modifiable graph can exist at one time which would be incompatible with this idea.

Reply:
> In that case it may be better to associate each parameter with an ID that can be used when setting parameters rather than requiring the user to maintain an up-to-date record of which pointer value was previously associated with the graph.

Yes, that makes sense.

> The idea of batching individual changes into an update function could be of benefit, though I don't believe it should return a new executable graph.

This seems fine too.

Reply:
This conversation seems to have stalled.  The description of `command_graph::update` still does not make it clear what "topologically identical" means precisely, so I think it will be hard for users to know what can be updated vs. what cannot.  I still think it would be clearer to instead have and API like `change_parameter` that made it more clear about what things the user can change in a finalized graph.

Reply:
Agree, we need to improve the definition of "topologically identical".
We do have a draft spec PR of an `change_parameter` like API - https://github.com/reble/llvm/pull/77 - but it would in addition to `command_graph::update` rather than instead of it.